packages/flutter/lib/src/painting/matrix_utils.dart

```diff
@@ -129,23 +129,98 @@ class MatrixUtils {
     return Offset(transformed3.x, transformed3.y);
   }
 
+  // Used to store temporary matrices in [transformRect].
+  static final Matrix4 _temporaryPointMatrix = Matrix4.zero();
+
   /// Returns a rect that bounds the result of applying the given matrix as a
   /// perspective transform to the given rect.
   ///
   /// This function assumes the given rect is in the plane with z equals 0.0.
   /// The transformed rect is then projected back into the plane with z equals
   /// 0.0 before computing its bounding rect.
   static Rect transformRect(Matrix4 transform, Rect rect) {
-    final Offset point1 = transformPoint(transform, rect.topLeft);
-    final Offset point2 = transformPoint(transform, rect.topRight);
-    final Offset point3 = transformPoint(transform, rect.bottomLeft);
-    final Offset point4 = transformPoint(transform, rect.bottomRight);
-    return Rect.fromLTRB(
-        _min4(point1.dx, point2.dx, point3.dx, point4.dx),
-        _min4(point1.dy, point2.dy, point3.dy, point4.dy),
-        _max4(point1.dx, point2.dx, point3.dx, point4.dx),
-        _max4(point1.dy, point2.dy, point3.dy, point4.dy),
-    );
+    // Put the coordinates of the four corners of the rect in a matrix in a
```

Comment:
I would do the following:

bx, by = transform(0, 0)
wx, wy = deltaTransform(0, w)
hx, hy = deltaTransform(h, 0)

this should be far fewer computations than either a matrix multiply or 4 vector multiplies.

The 4 points are now (bx, by), (bx+wx, by+wy), (bx+wx+hx, by+wy+hy), (bx+hx, by+hy)

You don't need to do all those additions to compute the bounding box, though, just do some min/maxing on the wxy,hxy values.

Reply:
What are `rectx`, `recty` and `rectw`?

Is `transform(rectx, recty)` the same as `transform.transform(Vector4(rectx, recty, 0, 1))`?

I'm also not familiar with `deltaTransform`. What does it do?

Reply:
The above optimization definitely works in the absence of perspective which I'm guessing is 99.999% of the operations we do, so even if we just test for a perspective transform and punt to the old code, the delta transform code will be much faster when they aren't performing 3D transforms.

Reply:
rectx = left
recty = top
rectw = right-left
recth = bottom-top

And yes, transform(x,y) is my shorthand for transform(x, y, 0, 1).

Basically in the absence of perspective the 4 points form a parallelogram (they start as a rectangle, but after transformation can be an arbitrary parallelogram) which can be transformed with 1 full transform and 2 delta transforms to save a lot of calculations.